src/psyclone/tests/nemo/test_files/implicit_do_structures.f90

```diff
@@ -36,14 +36,32 @@ program implicit_do_structures
   implicit none
   integer, parameter :: jpi=10, jpj=10, jpk=10
   real(kind=kind(1.0d0)), dimension(jpi,jpj,jpk) :: umask
+  integer, dimension(jpj) :: indices = 3
 
-  ! Test code with implicit NEMO-style do loop with structures
+  ! Test code with implicit do loop with structures
   umask(:,:,:) = mystruct%field(:,:,:) + mystruct%field2%field(:,:,:)
 
-  ! Test code with implicit NEMO-style do loop with structures in the LHS
+  ! Test code with implicit do loop with structures in the LHS
   mystruct%field2%field(:,:,:) = 0.0d0
 
-  ! Test code with implicit NEMO-style do loop with nested structures
-  umask(:,:,:) = mystruct%field(mystruct%field2%field3(:),:,:)
+  ! Test code with implicit do loop with AoSoA in the LHS
+  mystruct%field3(:,:,:)%field4 = 0.0d0
+
+  ! Test code mixing AoSoA in the LHS and SoA in the RHS
+  mystruct%field3(:,:,:)%field4 = mystruct%field2%field(:,:,:)
+
+  ! Test code with implicit do loop with 2 arrays in the LHS structure, one
+  ! without ranges
+  mystruct%field2(4, 3)%field(:,:,:) = mystruct%field2(5, 8)%field(:,:,:)
+
+  ! Test code with implicit do loop with 2 arrays in the LHS structure, both
+  ! with ranges
+  mystruct%field2(:,:)%field(:,:,:) = 0.0d0
```

Comment:
As discussed elsewhere, this isn't valid Fortran.

Reply:
I do agree that this is invalid Fortran and some frontend (fparser or fparserreader) should prevent it, so I deleted the Fortran code. I am not sure if this can never happen if somebody misconstructs PSyIR, so I misconstructed a tree for this test. I think this should stay at least until the tree misconstruction fails by itself.